emma-language/src/main/scala/org/emmalanguage/compiler/lang/source/Source.scala

```diff
@@ -157,6 +159,101 @@ trait Source extends Common
       //@formatter:on
     }
 
+    trait Algebra[A] {
+
+      // Empty tree
+      def empty: A
+
+      // Atomics
+      def lit(value: Any): A
+      def this_(encl: u.Symbol): A
+      def ref(target: u.TermSymbol): A
+
+      // References (with defaults)
+      def moduleRef(target: u.ModuleSymbol): A = ref(target)
+      def bindingRef(target: u.TermSymbol): A = ref(target)
+      def valRef(target: u.TermSymbol): A = bindingRef(target)
+      def varRef(target: u.TermSymbol): A = bindingRef(target)
+      def parRef(target: u.TermSymbol): A = bindingRef(target)
+
+      // Definitions
```

Comment:
Yes, DML supports recursion. Would it be possible to restrict the Source language using the validation methods? If we include it in the fold and Algebra you could then make sure that the validation catches it?